### PR TITLE
Prepare for finalizeDecl() removal

### PR DESCRIPTION
**Pull request by slavapestov**
An assortment of somewhat random refactorings.

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please test compiler performance

File: lib/Serialization/Serialization.cpp [hunk 1]
```diff
@@ -3499,6 +3500,16 @@ class Serializer::DeclSerializer : public DeclVisitor<DeclSerializer> {
   }
 
   void visitAccessorDecl(const AccessorDecl *fn) {
+    // Accessor synthesis and type checking is now sufficiently lazy that
+    // we might have unvalidated accessors in a primary file.
+    //
+    // FIXME: Once accessor synthesis and getInterfaceType() itself are
+    // request-ified this goes away.
+    if (!fn->hasValidSignature()) {
+      S.M->getASTContext().getLazyResolver()->resolveDeclSignature(
```
Comment (jrose-apple) [inline lib/Serialization/Serialization.cpp]: Can you assert that the accessor is synthesized in this case? We definitely do not want any explicitly-written accessors to be skipped.

File: lib/Sema/TypeCheckCircularity.cpp [hunk 2]
```diff
@@ -202,12 +199,6 @@ void CircularityChecker::run() {
       return;
     }
   }
-
-  // If we didn't report an error, but we encountered a property that
-  // hadn't been type-checked, queue the type for delayed checking.
-  if (RequireDelayedChecking) {
-    TC.DelayedCircularityChecks.push_back(OriginalDecl);
```
Comment (DougGregor) [inline lib/Sema/TypeCheckCircularity.cpp]: w00t!

File: lib/Serialization/Serialization.cpp (hunk 1 above)
Comment (slavapestov) [reply lib/Serialization/Serialization.cpp]: No problem. However do keep in mind this whole 'if' check will go away at some point.

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Pull request merged**